aioelasticsearch/helpers.py

```diff
@@ -147,3 +149,121 @@ def _update_state(self, resp):
         self._successful_shards = resp['_shards']['successful']
         self._total_shards = resp['_shards']['total']
         self._done = not self._hits or self._scroll_id is None
+
+
+async def worker_bulk(client, datas , actions,  **kwargs):
+    try:
+        resp = await client.bulk("\n".join(actions) + '\n', **kwargs)
+    except TransportError as e:
+        return e, datas
+    fail_actions = []
+    finish_count = 0
+    for data, (op_type, item) in zip(datas, map(methodcaller('popitem'),
+                                                resp['items'])):
+        ok = 200 <= item.get('status', 500) < 300
+        if not ok:
+            fail_actions.append(data)
+        else:
+            finish_count += 1
+    return finish_count, fail_actions
+
+
+def _get_fail_data(results, serializer):
+    finish_count = 0
+    bulk_action = []
+    bulk_data = []
+    lazy_exception = None
+    for result in results:
+        if isinstance(result[0], int):
+            finish_count += result[0]
+        else:
+            if lazy_exception is None:
+                lazy_exception = result[0]
+
+        for fail_data in result[1]:
+            for _ in fail_data:
+                bulk_data.append(_)
+        if result[1]:
+            bulk_action.extend(map(serializer.dumps,result[1]))
+    return finish_count, bulk_data, bulk_action, lazy_exception
+
+
+async def _retry_handler(client, futures, max_retries, initial_backoff,
+                         max_backoff, **kwargs):
+    finish = 0
+    for attempt in range(max_retries + 1):
+        if attempt:
+            sleep = min(max_backoff, initial_backoff * 2 ** (attempt - 1))
+            await asyncio.sleep(sleep)
+
+        results = await asyncio.gather(*futures,
```

Comment:
This gather produces unexpected load, not sure, but maybe it should be rewritten in `asyncio.queue` / workers

Reply:
provide please explicit loop